### PR TITLE
WASM fixes

### PR DESCRIPTION
**Pull request by uncomputable**
Fixes issues that came up while parsing the human encoding in WASM.

### CODE REVIEW

**Comment by apoelstra** [comment]
`thread_local` is more efficient than `Once` because it doesn't need to use synchronization every time that it's accessed.

**Comment by apoelstra** [comment]
If WASM can't handle `thread_local`, even though I wrote the original code specifically with WASM in mind (in which it should be a no-op because WASM has no threads), then I guess it's fine to hack around it by making our code less efficient and unsafe. But I'm not thrilled about it.

**Comment by apoelstra** [comment]
I can't test this until we update the vendored version of Simplicity for the recent CI fixes.

File: src/human_encoding/parse/mod.rs [hunk 1]
```diff
@@ -749,13 +749,29 @@ mod tests {
...
+            ("0xf", Value::u4(0xf)),
+            ("0x00", Value::u8(0x00)),
+            ("0xff", Value::u8(0xff)),
+            ("0xdeadbeef", Value::power_of_two(&[0xde, 0xad, 0xbe, 0xef])),
+        ];
+
+        for (human, value) in human_values {
+            let s = format!("main := comp const {human} unit");
+            assert_const::<Core>(s.as_str(), value);
+        }
```
Comment (apoelstra) [inline src/human_encoding/parse/mod.rs:775]: In 8355454f8efb4006c0df2c587c6dd19a6cb5857b:

This test fails on my system with 
```
failures:

---- human_encoding::parse::tests::const_word stdout ----
thread 'human_encoding::parse::tests::const_word' panicked at 'byte index 2 is out of bounds of `0`', src/human_encoding/parse/ast.rs:365:47
stack backtrace:
   0: rust_begin_unwind
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panicking.rs:67:14
   2: core::str::slice_error_fail_rt
   3: core::str::slice_error_fail
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/str/mod.rs:87:9
   4: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::Range<usize>>::index
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/str/traits.rs:235:21
   5: core::str::traits::<impl core::ops::index::Index<I> for str>::index
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/str/traits.rs:61:9
   6: simplicity::human_encoding::parse::ast::Ast<J>::from_literal_lexeme
             at ./src/human_encoding/parse/ast.rs:365:47
   7: core::ops::function::Fn::call
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/ops/function.rs:79:5
   8: santiago::parser::tree::Tree<AST>::as_abstract_syntax_tree
             at /store/home/apoelstra/.cargo/registry/src/index.crates.io-6f17d22bba15001f/santiago-1.3.1/src/parser/tree.rs:91:46
   9: simplicity::human_encoding::parse::ast::parse_line_vector
             at ./src/human_encoding/parse/ast.rs:122:11
  10: simplicity::human_encoding::parse::parse
             at ./src/human_encoding/parse/mod.rs:180:17
  11: simplicity::human_encoding::parse::tests::assert_const
             at ./src/human_encoding/parse/mod.rs:642:15
  12: simplicity::human_encoding::parse::tests::const_word
             at ./src/human_encoding/parse/mod.rs:774:13
  13: simplicity::human_encoding::parse::tests::const_word::{{closure}}
             at ./src/human_encoding/parse/mod.rs:751:21
  14: core::ops::function::FnOnce::call_once
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/ops/function.rs:250:5
  15: core::ops::function::FnOnce::call_once
             at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I think you just need to reorder the commit that introduces tests with the one that fixes parsing of single-character hex literals.

Comment (uncomputable) [reply src/human_encoding/parse/mod.rs:775]: Yes, I will reorder the commits. I will also try to fix the precomputed powers of two without using `Once`.

**Comment by uncomputable** [comment]
For now, I reordered the commits to make the tests work.

**Review by apoelstra — APPROVED** [review]
ACK 97e41f540c6c1b6234ef1cb219870f2d5c08c227

**Comment by uncomputable** [comment]
`thread_local!` works, so I added it back in. The borrow error no longer occurs.

We could share the same powers of two across threads using `RwLock` or `OnceLock`, but this requires MSRV 1.63.0 and MSRV 1.70.0, respectively, to work in `const`.

**Comment by apoelstra** [comment]
`RwLock` is extremely slow and should basically never be used unless your problem architecture really requires it. Not sure about `OnceLock`. It's likely to be better.

**Review by apoelstra — APPROVED** [review]
ACK c57d8def80cffb360a667101fd36990a83cee969

**Pull request merged**